### PR TITLE
crd: Add load distribution policy crd and generated client code

### PR DESCRIPTION
**Pull request by erickhernandezascencio**
## Checklist
- [X] The title of the PR is clear and informative
- [ ] If applicable, the changes made in the PR have proper test coverage
- [ ] Issues addressed by the PR are mentioned in the description followed by `Fixes`.

## Description

This PR includes the newly generated files for the LDP backend CRD

### CODE REVIEW

File: crds/AzureApplicationGatewayLoadDistributionPolicy.yaml [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+                      backend:
+                        type: object
+                        properties:
+                          service:
+                            description: "service object model"
+                            type: object
+                            properties:
+                              name:
+                                type: string
+                                description: "same of service"
```
Comment (akshaysngupta) [inline crds/AzureApplicationGatewayLoadDistributionPolicy.yaml:40]: `Name is the referenced service. The service must exist in the same namespace as the Ingress object.`

File: crds/AzureApplicationGatewayLoadDistributionPolicy.yaml [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+                  description: "An array of services that contains information for how the service should be used for load balancing and active passive support in Application Gateway"
+                  type: array
+                  items:
+                    type: object
+                    properties:
+                      backend:
+                        type: object
+                        properties:
+                          service:
+                            description: "service object model"
```
Comment (akshaysngupta) [inline crds/AzureApplicationGatewayLoadDistributionPolicy.yaml:35]: service to load balancer

File: crds/examples/AzureApplicationGatewayLoadDistributionPolicy.yaml [hunk 3]
```diff
@@ -0,0 +1,20 @@
+apiVersion: appgw.ingress.azure.io/v1
+kind: AzureApplicationGatewayLoadDistribitionPolicy
+metadata:
+  name: load-distribution-policy-with-failover
+spec:
+  targets:
+    - backend:
+        service:
+          name: gwm-mcs-prod
```
Comment (akshaysngupta) [inline crds/examples/AzureApplicationGatewayLoadDistributionPolicy.yaml:9]: service-v1

File: crds/examples/AzureApplicationGatewayLoadDistributionPolicy.yaml [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+          port:
+            number: 80
+      role: active
+      weight: 8
+    - backend:
+        service:
+          name: gwm-mcs-staging
```
Comment (akshaysngupta) [inline crds/examples/AzureApplicationGatewayLoadDistributionPolicy.yaml:16]: service-v2

File: crds/AzureApplicationGatewayLoadDistributionPolicy.yaml [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+  group: appgw.ingress.azure.io
+  scope: Namespaced
+  names:
+    kind: AzureApplicationGatewayLoadDistributionPolicy
+    plural: azureapplicationgatewayloaddistributionpolicies
+    singular: azureapplicationgatewayloaddistributionpolicy
+    shortNames:
+      - agldp
+  versions:
+    - name: v1
```
Comment (akshaysngupta) [inline crds/AzureApplicationGatewayLoadDistributionPolicy.yaml:15]: v1beta1

File: crds/AzureApplicationGatewayLoadDistributionPolicy.yaml [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+      storage: true
+      schema:
+        openAPIV3Schema:
+          type: object
+          properties:
+            spec:
+              type: object
+              properties:
+                targets:
+                  description: "An array of services that contains information for how the service should be used for load balancing and active passive support in Application Gateway"
```
Comment (akshaysngupta) [inline crds/AzureApplicationGatewayLoadDistributionPolicy.yaml:26]: `An array of services that contains information for how the service should be used for load balancing in Application Gateway`

File: crds/AzureApplicationGatewayLoadDistributionPolicy.yaml (hunk 2 above)
Comment (akshaysngupta) [reply crds/AzureApplicationGatewayLoadDistributionPolicy.yaml:35]: `Service references a Service as a Backend.`

File: crds/AzureApplicationGatewayLoadDistributionPolicy.yaml [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+                        properties:
+                          service:
+                            description: "service object model"
+                            type: object
+                            properties:
+                              name:
+                                type: string
+                                description: "same of service"
+                              port:
+                                type: object
```
Comment (akshaysngupta) [inline crds/AzureApplicationGatewayLoadDistributionPolicy.yaml:42]: description for Port: `Port of the referenced service. A port name or port number`

Comment (akshaysngupta) [reply crds/AzureApplicationGatewayLoadDistributionPolicy.yaml:42]: you can reference service related description from the v1.Ingress API definition

https://github.com/kubernetes/api/blob/master/networking/v1/types.go#L438

File: crds/AzureApplicationGatewayLoadDistributionPolicy.yaml [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+                                properties:
+                                  number:
+                                    type: integer
+                                    description: "port number for service"
+                                  name:
+                                    type: string
+                                    decription: "port name for service"
+                          weight:
+                            description: "weight of the service when Application Gateway uses it for weighted round robin load balancing"
```
Comment (akshaysngupta) [inline crds/AzureApplicationGatewayLoadDistributionPolicy.yaml:51]: `Weight specifies the proportion of HTTP requests forwarded to the backend.`

File: crds/AzureApplicationGatewayLoadDistributionPolicy.yaml [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+                            type: integer
+                            default: 1
+                          role:
+                            description: "Decides if server should be active or passive for Active/Passive load balancing"
```
Comment (akshaysngupta) [inline crds/AzureApplicationGatewayLoadDistributionPolicy.yaml:55]: Specifies whether a server should be active or passive. Passive backends will receive HTTP requests only when all active backends are marked unhealthy by the Application Gateway's heath probe.

**Review by akshaysngupta — CHANGES_REQUESTED** [review]
Added comments

File: crds/AzureApplicationGatewayLoadDistributionPolicy.yaml (hunk 8 above)
Comment (akshaysngupta) [reply crds/AzureApplicationGatewayLoadDistributionPolicy.yaml:51]: referenced: https://gateway-api.sigs.k8s.io/references/spec/#networking.x-k8s.io/v1alpha1.HTTPRouteForwardTo

**Review by akshaysngupta — DISMISSED** [review]
LGTM

**Review by akshaysngupta — DISMISSED** [review]
LGTM

File: crds/examples/AzureApplicationGatewayLoadDistributionPolicy.yaml [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+          port:
+            number: 80
+      role: active
+      weight: 8
+    - backend:
+        service:
+          name: service-2
+          port:
+            number: 80
+      role: active
```
Comment (3quanfeng) [inline crds/examples/AzureApplicationGatewayLoadDistributionPolicy.yaml:19]: passive?

File: crds/examples/AzureApplicationGatewayLoadDistributionPolicy.yaml [hunk 11]
```diff
@@ -0,0 +1,20 @@
+apiVersion: appgw.ingress.azure.io/v1beta1
+kind: AzureApplicationGatewayLoadDistribitionPolicy
+metadata:
+  name: load-distribution-policy-with-failover
```
Comment (3quanfeng) [inline crds/examples/AzureApplicationGatewayLoadDistributionPolicy.yaml:4]: Does it have to be active-passive pair?

File: crds/AzureApplicationGatewayLoadDistributionPolicy.yaml [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+                                  number:
+                                    type: integer
+                                  name:
+                                    type: string
+                          weight:
+                            description: "Weight specifies the proportion of HTTP requests forwarded to the backend."
+                            type: integer
+                            default: 1
+                          role:
+                            description: "Specifies whether a server should be active or passive. Passive backends will receive HTTP requests only when all active backends are marked unhealthy by the Application Gateway's heath probe."
```
[the remark this answers is not in the document]
Comment (erickhernandezascencio) [reply crds/AzureApplicationGatewayLoadDistributionPolicy.yaml:54]: no, currently the set up is that only if all active services are down then start serving traffic to the passive pools

File: crds/examples/AzureApplicationGatewayLoadDistributionPolicy.yaml (hunk 10 above)
Comment (erickhernandezascencio) [reply crds/examples/AzureApplicationGatewayLoadDistributionPolicy.yaml:19]: correct, this should be passive to be more in line with example. will fix

File: crds/examples/AzureApplicationGatewayLoadDistributionPolicy.yaml (hunk 11 above)
Comment (erickhernandezascencio) [reply crds/examples/AzureApplicationGatewayLoadDistributionPolicy.yaml:4]: it does not have to be. someone can have all active servers if they would like. services are grouped by what LDP they belong to

File: crds/AzureApplicationGatewayLoadDistributionPolicy.yaml [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+                                description: "Name is the referenced service. The service must exist in the same namespace as the Ingress object."
+                              port:
+                                type: object
+                                description: "Port of the referenced service. A port name or port number"
+                                properties:
+                                  number:
+                                    type: integer
+                                  name:
+                                    type: string
+                          weight:
```
Comment (akshaysngupta) [inline crds/AzureApplicationGatewayLoadDistributionPolicy.yaml:49]: Suggestion: how about we shift `weight` and `role` outside `backend` property
```
    -  weight: 2
       role: passive
       backend:
        service:
          name: new-service
          port:
            number: 80
    -  weight: 8
       role: active
       backend:
        service:
          name: old-service
          port:
            number: 80
```

File: pkg/apis/azureapplicationgatewayloaddistributionpolicy/v1beta1/types.go [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+
+// AzureApplicationGatewayLoadDistributionPolicySpec defines a list of backend pool addresses
+type AzureApplicationGatewayLoadDistributionPolicySpec struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	//Targets include a list of backend targets
+	Targets []Backend `json:"Targets,omitempty"`
```
Comment (akshaysngupta) [inline pkg/apis/azureapplicationgatewayloaddistributionpolicy/v1beta1/types.go:34]: replace with
```go
Targets []Target `json:"Targets,omitempty"
```

File: pkg/apis/azureapplicationgatewayloaddistributionpolicy/v1beta1/types.go [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+}
+
+// Backend defines a backend service and its load distribution policy parameters
+type Backend struct {
```
Comment (akshaysngupta) [inline pkg/apis/azureapplicationgatewayloaddistributionpolicy/v1beta1/types.go:38]: Replace with:
```go
type Target struct {
	Role    string                   `json:"role,omitempty"`
	Weight  int                      `json:"weight,omitempty"`
	Backend Backend           `json:"backend,omitempty"`
}
```

Comment (akshaysngupta) [reply pkg/apis/azureapplicationgatewayloaddistributionpolicy/v1beta1/types.go:38]: similar to https://github.com/kubernetes/api/blob/master/networking/v1/types.go#L433

File: crds/AzureApplicationGatewayLoadDistributionPolicy.yaml [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+                                description: "Port of the referenced service. A port name or port number"
+                                properties:
+                                  number:
+                                    type: integer
+                                  name:
+                                    type: string
+                      weight:
+                        description: "Weight specifies the proportion of HTTP requests forwarded to the backend."
+                        type: integer
+                        default: 1
```
Comment (akshaysngupta) [inline crds/AzureApplicationGatewayLoadDistributionPolicy.yaml:52]: Add min and max:
```
minimum: 1
maximum: 1000
```

https://swagger.io/docs/specification/data-models/data-types/#numbers

File: crds/AzureApplicationGatewayLoadDistributionPolicy.yaml [hunk 17]
```diff
@@ -0,0 +1,56 @@
...
+                      role:
+                        description: "Specifies whether a server should be active or passive. Passive backends will receive HTTP requests only when all active backends are marked unhealthy by the Application Gateway's heath probe."
+                        type: string
+                        default: "active"
```
Comment (akshaysngupta) [inline crds/AzureApplicationGatewayLoadDistributionPolicy.yaml:56]:
```yaml
pattern: "^active$|^passive$"
```

**Review by akshaysngupta — APPROVED** [review]
LGTM

**Pull request merged**